core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,166 @@
+                              Tag.of(SERVER_TAG, dnsServerAddress.getAddress().getHostAddress())))
+                              .increment();
+    }
+
+    @Override
+    public void queryCancelled(int queriesRemaining) {
+        meterRegistry.counter(meterIdPrefix.name() + ".cancelled",
+                              Tags.of(nameTag)).increment();
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryRedirected(List<InetSocketAddress> nameServers) {
+        meterRegistry.counter(meterIdPrefix.name() + ".redirected",
+                              Arrays.asList(nameTag, Tag.of(SERVERS_TAG,
+                              nameServers.stream().map(addr -> addr.getAddress().getHostAddress())
+                              .collect(Collectors.joining(","))))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryCNAMEd(DnsQuestion cnameQuestion) {
+        meterRegistry.counter(meterIdPrefix.name() + ".cnamed",
+                              Arrays.asList(nameTag,
+                              Tag.of(CNAME_TAG, cnameQuestion.name()))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryNoAnswer(DnsResponseCode code) {
+        meterRegistry.counter(meterIdPrefix.name() + ".noanswer",
+                              Arrays.asList(nameTag,
+                              Tag.of(CODE_TAG, String.valueOf(code.intValue())))).increment();
+        return this;
+    }
+
+    @Override
+    public void queryFailed(Throwable cause) {
+        meterRegistry.counter(meterIdPrefix.name(),
+                              Arrays.asList(nameTag, Tag.of(RESULT_TAG, "failure"),
+                              Tag.of(CAUSE_TAG, determineDnsExceptionTag(cause).lowerCasedName))).increment();
+    }
+
+    @Override
+    public void querySucceed() {
+        meterRegistry.counter(meterIdPrefix.name(),
+                              Arrays.asList(nameTag, Tag.of(RESULT_TAG, "success"),
+                              Tag.of(CAUSE_TAG, "none"))).increment();
+    }
+
+    private static DnsExceptionTypes determineDnsExceptionTag(Throwable cause) {
+        if (cause instanceof DnsTimeoutException) {
+            return DnsExceptionTypes.SERVER_TIMEOUT;
+        } else if (cause instanceof DnsNameResolverTimeoutException) {
+            return DnsExceptionTypes.RESOLVER_TIMEOUT;
+        }
+        final String message = cause.getMessage();
+        if (NXDOMAIN_EXCEPTION.matcher(message).find()) {
+            return DnsExceptionTypes.NX_DOMAIN;
+        }
+
+        if (CNAME_EXCEPTION.matcher(message).find()) {
+            return DnsExceptionTypes.CNAME_NOT_FOUND;
+        }
+
+        if (NO_MATCHING_EXCEPTION.matcher(message).find()) {
+            return DnsExceptionTypes.NO_MATCHING_RECORD;
+        }
+
+        if (UNRECOGNIZED_TYPE_EXCEPTION.matcher(message).find()) {
+            return DnsExceptionTypes.UNRECOGNIZED_TYPE;
+        }
+
```

Comment:
Shouldn't we ignore the 'exhausted name servers' error here instead of returning `OTHERS`, so that the counter doesn't increase?

Reply:
Are you suggesting exhausted name servers exception should not be tagged in a counter that is incremented queryFailed callback ? 
I am also noticing this exception is being thrown "No name servers returned an answer" but we capture it as other. Don't know what we want to do here.

Reply:
> Are you suggesting exhausted name servers exception should not be tagged in a counter that is incremented queryFailed callback ?

Yes, but see below:

> I am also noticing this exception is being thrown "No name servers returned an answer" but we capture it as other. Don't know what we want to do here.

That's a good point. Is it possible to count "No name servers returned an answer" as a separate counter, and ignore other 'exhausted name servers' exceptions?

Reply:
Done.

Reply:
@trustin Sorry for the confusion the "No name servers returned an answer" is basically NAME_SERVERS_EXHAUSTED_EXCEPTION returned by the DnsResolveContext[netty class] that we had agreed to ignore.